lib/recursive-methods.rb

```diff
@@ -1,49 +1,86 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) 
+# Space complexity: O(n) 
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError.new if n < 0 
+  return 1 if n == 1 || n == 0  
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) 
+# Space complexity: O(n)  
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  reversed_s = reverse(s[1..-1])
```

Comment:
`s[1..-1]` creates a new array and copies all the elements over into the new array... So this line by itself is O(n)